ecleankernel/file.py

```diff
@@ -226,6 +227,17 @@ def read_version_from_efi(self,
             buf = f.read(0x40)
             if len(buf) != 0x40 or buf[:2] != b"MZ":
                 return None
+
+            # handle EFI zboot image
+            # see kernel source code drivers/firmware/efi/libstub/zboot-header.S
+            if buf[4:8] == b"zimg":
+                offset, size = struct.unpack_from("<LL", buf, 8)
+
+                f.seek(offset)
+                image = f.read(size)
+
+                return self.read_version_from_raw(io.BytesIO(image))
```

Comment:
Any reason you're prereading it, rather than calling `self.read_version_from_raw` on the seeked file?

Reply:
`gzip.compress` throws when it tries to read past the end of the payload.

Reply:
Can we instead modify `read_version_from_raw()` and related functions to take an optional `size`?